### PR TITLE
Project custom rdfs labels

### LINKED ISSUE #98: RDF Serializer > Project Custom rdfs:label
# Topology: ProjectCustomRdfsLabels

This topology produces custom rdfs:label for all projects, that are needed to properly display labels of auto generated
classes and properties (since their labels are not provided by ontome).

```mermaid
flowchart TD
    1a-->2a-->3a-->3b
 
    subgraph 1
        1a([1a project])
    end
    subgraph __2
        2a([2a FlatMap])
    end          
    subgraph __3
        3a([3a To])
        3b[3b project_rdf]
    end
    
```

1a) Create a KStream from `{dev/stag/prod}.projects.project`

2a) FlatMap the records to `ProjectRdfKey` and `ProjectRdfValue` to these triples:

```turtle
<http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/2000/01/rdf-schema#label> "has type"@en .
<http://www.w3.org/2000/01/rdf-schema#label> <http://www.w3.org/2000/01/rdf-schema#label> "has label"@en .
<http://www.w3.org/2002/07/owl#sameAs> <http://www.w3.org/2000/01/rdf-schema#label> "same as"@en . 

<https://ontome.net/ontology/c50> <http://www.w3.org/2000/01/rdf-schema#label> "Time-Span"@en .
<https://ontome.net/ontology/c50> a <http://www.w3.org/2002/07/owl#Class> .
<https://ontome.net/ontology/p4> <http://www.w3.org/2000/01/rdf-schema#label> "has time-span"@en .
<https://ontome.net/ontology/p4i> <http://www.w3.org/2000/01/rdf-schema#label> "is time-span of"@en .

<http://www.w3.org/2006/time#DateTimeDescription> <http://www.w3.org/2000/01/rdf-schema#label> "Date-Time description"@en .
<http://www.w3.org/2006/time#hasTRS> <http://www.w3.org/2000/01/rdf-schema#label> "Temporal reference system used"@en .
<http://www.w3.org/2006/time#unitType> <http://www.w3.org/2000/01/rdf-schema#label> "temporal unit type"@en .
<http://www.w3.org/2006/time#year> <http://www.w3.org/2000/01/rdf-schema#label> "Year"@en .
<http://www.w3.org/2006/time#month> <http://www.w3.org/2000/01/rdf-schema#label> "Month"@en .
<http://www.w3.org/2006/time#day> <http://www.w3.org/2000/01/rdf-schema#label> "Day"@en .
<http://www.w3.org/2006/time#unitYear> <http://www.w3.org/2000/01/rdf-schema#label> "Year (unit of temporal duration)"@en .
<http://www.w3.org/2006/time#unitMonth> <http://www.w3.org/2000/01/rdf-schema#label> "Month (unit of temporal duration)"@en .
<http://www.w3.org/2006/time#unitDay> <http://www.w3.org/2000/01/rdf-schema#label> "Day (unit of temporal duration)"@en .

<http://www.opengis.net/def/uom/ISO-8601/0/Gregorian> <http://www.w3.org/2000/01/rdf-schema#label> "Gregorian Calendar"@en .
<https://d-nb.info/gnd/4318310-4> <http://www.w3.org/2000/01/rdf-schema#label> "Julian Calendar"@en .
```

3a) To: sink it to `project_rdf`

### PR DESCRIPTION
**Pull request by dferhod**
resolves #98

### CODE REVIEW

**Comment by dferhod** [comment]
@joschne please review this PR :)

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public void addProcessorsStandalone() {
+         addProcessors(
+                registerInputTopic.projectStream()
+        );
+    }
+
+    public ProjectRdfReturnValue addProcessors(
+            KStream<ProjectRdfKey, ProjectRdfValue> projectStream
```
Comment (joschne) [inline app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java:47]: The input stream is KStream<dev.projects.project.Key, dev.projects.project.Value>

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/RegisterInputTopic.java [hunk 2]
```diff
@@ -66,4 +66,13 @@ public KStream<ProjectEntityKey, ProjectEntityLabelValue> projectEntityLabelStre
         );
     }
 
+    public KStream<ProjectRdfKey, ProjectRdfValue> projectStream() {
```
Comment (joschne) [inline app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/RegisterInputTopic.java:69]: The input stream is KStream<dev.projects.project.Key, dev.projects.project.Value>

Also update the serdes below.

File: app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+import java.util.Properties;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ProjectCustomRdfsLabelsTest {
+
+    private static final String SCHEMA_REGISTRY_SCOPE = ProjectCustomRdfsLabelsTest.class.getName();
+    private static final String MOCK_SCHEMA_REGISTRY_URL = "mock://" + SCHEMA_REGISTRY_SCOPE;
+    private TopologyTestDriver testDriver;
+    private TestInputTopic<ProjectRdfKey, ProjectRdfValue> projectTopic;
```
Comment (joschne) [inline app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java:25]: The input stream is KStream<dev.projects.project.Key, dev.projects.project.Value>

**Review by joschne — CHANGES_REQUESTED** [review]
Requested changes: The key and value schemas of the input topic have to be changed. Else it looks good.

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+                    turtles.add("<" + GREG.getUrl() + "> <" + RDFS.getUrl() + "label> \"Gregorian Calendar\"@en .");
+                    turtles.add("<" + JUL.getUrl() + "> <" + RDFS.getUrl() + "label> \"Julian Calendar\"@en .");
+
+
+
+                    ProjectRdfKey k;
+                    for (String item : turtles) {
+                        k = ProjectRdfKey.newBuilder()
+                                .setProjectId(key.getProjectId())
```
Comment (joschne) [inline app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java:90]: projectId is key.pk_entity

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java (hunk 1 above)
Comment (dferhod) [reply app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java]: Done

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/RegisterInputTopic.java (hunk 2 above)
Comment (dferhod) [reply app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/RegisterInputTopic.java]: Done

File: app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java (hunk 3 above)
Comment (dferhod) [reply app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java]: Done

File: app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java (hunk 4 above)
Comment (dferhod) [reply app-5-rdf-serializer/src/main/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabels.java]: Done

**Comment by dferhod** [comment]
Changes requested done.
Please review again @joschne :)

File: app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        var projectId = 1;
+
+        var k = dev.projects.project.Key.newBuilder()
+                .setPkEntity(projectId)
+                .build();
+        var v = dev.projects.project.Value.newBuilder()
+                .setFkLanguage(I.EN.get())
+                .build();
+
+        /*var k = ProjectRdfKey.newBuilder().setProjectId(projectId).setTurtle("").build();
```
Comment (joschne) [inline app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java:76]: please remove commented section

**Review by joschne — CHANGES_REQUESTED** [review]
I only suggest to remove commented code. No additional review needed after this change.

File: app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java (hunk 5 above)
Comment (dferhod) [reply app-5-rdf-serializer/src/test/java/org/geovistory/toolbox/streams/rdf/processors/project/ProjectCustomRdfsLabelsTest.java]: Done

**Pull request merged**